openedx_events/event_bus/__init__.py

```diff
@@ -116,11 +119,67 @@ def get_producer() -> EventBusProducer:
     If misconfigured, returns a fake implementation that can be called but does nothing.
     """
     return _try_load(
-        setting_name='EVENT_BUS_PRODUCER',
+        setting_name='EVENT_BUS_PRODUCER', args=(), kwargs={},
         expected_class=EventBusProducer, default=NoEventBusProducer(),
     )
 
 
+class EventBusConsumer(ABC):
+    """
+    Parent class for event bus consumer implementations.
+    """
+
+    @abstractmethod
+    def consume_indefinitely(self) -> None:
+        """
+        Consume events from a topic in an infinite loop.
+
+        Events will be converted into calls to Django signals.
+        """
+
+
+class NoEventBusConsumer(EventBusConsumer):
+    """
+    Stub implementation to "load" when no implementation is properly configured.
+    """
+
+    def consume_indefinitely(self) -> None:
+        """Do nothing."""
+
+
+# .. setting_name: EVENT_BUS_CONSUMER
+# .. setting_default: None
+# .. setting_description: String naming a callable (function or class) that can be called to create
+#   or retrieve an instance of EventBusConsumer when ``openedx_events.event_bus.make_single_consumer`` is
+#   called. The format of the string is a dotted path to an attribute in a module, e.g.
+#   ``some.module.path.EventBusImplementation``. See docstring of ``make_single_consumer`` for
+#   parameters. If setting is not supplied or the callable raises an exception or does not return
+#   an instance of EventBusConsumer, calls to the consumer will be ignored with a warning at startup.
+
+
+def make_single_consumer(*, topic: str, group_id: str, signal: OpenEdxPublicSignal, **kwargs) -> EventBusConsumer:
```

Comment:
These kwargs are here to provide additional implementation-specific configuration, right? In the Kafka implementation we just used Django settings namespaced under `EVENT_BUS_KAFKA_` for that purpose, and I would prefer that if possible. Here's an example of that:

- `create_producer` reads configuration before creating an `EventBusProducer`: https://github.com/openedx/event-bus-kafka/blob/main/edx_event_bus_kafka/internal/producer.py#L339
- Configuration is loaded from Django settings: https://github.com/openedx/event-bus-kafka/blob/main/edx_event_bus_kafka/internal/config.py#L72

Is there some other constraint that makes this JSON-and-kwargs preferable here?

Reply:
@timmc-edx The reason was that the values for kwargs can change case by case and they are meant to be used in  special cases only, for example, if the users want to process previous events from a specific point or they want to run multiple consumers in a group so they need to pass in `--consumer_name` for each worker which should be unique in the group. Adding these in the settings means they will need to update the settings each time they want to start a worker.

Reply:
That's actually how we've been handling it with Kafka -- change the settings, which automatically causes a redeploy of the workers on kubernetes, and then change them back again. But I can see how that approach might not be desirable for all deployers.

Reply:
In our case, the Django settings and the command line arguments are in the same file as each other, so it doesn't really matter much which mechanism we use. I don't know how many people would be using this kind of setup, though.

If Redis needs individual names for the consumers in a group, I guess it seems more likely that you'd have a list of individual workers hardcoded into your infrastructure, each with its own command line. It might be harder to give each one its own Django settings. (That would be true for our infrastructure, in fact.)

So I'm fine with the `--extra` arg. (Well, I feel like there could be a better name for it, but I can't actually _come up with_ anything better, so I really can't complain there!)

Reply:
Oh! I was actually totally wrong about how we were passing offset time to Kafka. We're using a CLI argument to the management command, just like you noted in the PR description. :-) So I guess we'd make use of this as well!

Reply:
Cool, I'll leave this as is then.